src/hooks/chatSidebar.ts

```diff
@@ -38,6 +42,13 @@ export const useChatSidebarControls = () => {
     }
   }, [expandSidebar, collapseSidebar, isExpanded]);
 
+  useMousetrap({
+    keys: KeyboardShortcutKeys.chat,
+    callback: toggleSidebar,
```

Comment:
somehow this is not working for a case I have described in PR description: If go to "Direct messages" and then back to "Party chat" - "c" no longer toggles chat.  Going back to Direct or toggling chat with mouse makes everything work again.  @0xdevalias any ideas?

Reply:
I believe this issue was what I originally shared https://github.com/sparkletown/sparkle/pull/1416#discussion_r636484294 about.

Reply:
I haven't yet looked specifically into this and any deeper potential reasons beyond those original musing thoughts.

Reply:
Ok, so looking at the definition of `toggleSidebar` I expect that a 'stale function reference' is the most likely case here.

Whereas the other 'keybound togglers' use a function that never changes, `toggleSidebar` gets recreated whenever any of `expandSidebar, collapseSidebar, isExpanded` changes. So when you run it the first time, `isExpanded` will change, and so the function gets recreated. Yet `useMousetrap` likely has the old version of the function bound/cached still, so it just constantly tries to run the 'same' toggle (eg. open sidebar) every time.

Need to look a bit deeper to figure the easiest/best solution here.

Reply:
Looking at the definition of `useMousetrap`, it should unregister+re-register itself whenever the `useEffect`'s deps array changes, which includes the `callback` prop:

- https://github.com/sparkletown/sparkle/blob/staging/src/hooks/useMousetrap.ts#L70-L87

So I would have expected the change of `toggleSidebar` to cause this to be re-registered and work as expected..

---

To confirm/start debugging further from here, I would add a `console.log` within `useMousetrap`'s `useEffect` to see if it is being rebound when it should be; and I would also add a `console.log` to `toggleSidebar` to see if it's being called (and what the value of `isExapanded` is) each time you press the shortcut.

That should at least get more information to figure out whether it's following the 'static' expectations/assumptions i've made above from looking at the code.

Reply:
WONDERFUL explanation and pointers elsewhere -- even I managed to troubleshoot it to death... Added console.log first, saw that no mousetraps when switching back... looked at other *Chat*.ts* performed -- they all seems to have some kind of onClick which triggers specific to them select*Chat.  Added similar one to VenueChat - All is working now . Pushed 08c63ff6272e792a0ccb2474f95a600918a37611 .

Reply:
resolved now with 08c63ff6272e792a0ccb2474f95a600918a37611 + 8f02d07eb87046f6f2644eb1cc1a6c583bb42afa

Works good as to me.  Anything else or could be blessed? ;)

Reply:
> looked at other Chat.ts* performed -- they all seems to have some kind of onClick which triggers specific to them select*Chat. 

@yarikoptic When you say 'other Chat.ts', which components/etc are you referring to?

Looking at usages of `useChatSidebarControls` i'm guessing you mean these?:

- `src/components/organisms/ChatSidebar/ChatSidebar.tsx`
  - `const { isExpanded, toggleSidebar, chatSettings, selectVenueChat, selectPrivateChat } = useChatSidebarControls();`
- `src/components/organisms/ChatSidebar/components/PrivateChats/PrivateChats.tsx`
  - `const { selectRecipientChat } = useChatSidebarControls();`
    - note that this is `selectRecipientChat`, yet we are in `PrivateChats`
- `src/components/organisms/ChatSidebar/components/RecipientChat/RecipientChat.tsx`
  - `const { selectPrivateChat } = useChatSidebarControls();`
    - note that this is `selectPrivateChat`, yet we are in `RecipientChat`

Looking at the changes in 08c63ff + 8f02d07, you've added:

- `src/components/organisms/ChatSidebar/components/VenueChat/VenueChat.tsx`
  - `const { selectVenueChat } = useChatSidebarControls();`
    - note that this is `selectVenueChat`, and we are in `VenueChat`
    
While this might have made it work, if it did, it's likely as an unrelated side effect rather than this being the proper fix.

Without looking too deeply into the specifics of that code, I believe:

- `PrivateChats` uses `selectRecipientChat` to switch to the 'show me my private messages with this particular user' view
- `RecipientChat` uses `selectPrivateChat` to switch back to the 'show me a list of all users I have private chats with' view

I don't believe there is any valid reason to add `selectVenueChat` as an `onClick` in `VenueChat`.

Reply:
nice analysis! and yes -- it is those which I saw but missed that "pairing" pattern you point to.  It somewhat partially explains it though I guess -- since PrivateChats does have a call to `useChatSidebarControls` - it overloads them. When going to `venueChat`, no call to `useChatSidebarControls` and some things are stale, shortcuts do not work.  So it might be just a matter of having it bound just to trigger that code and ensure proper rebinds. I have looked around but see no better place to stick that onClick.    I have tested quite a bit and did not spot any oddity (besides it working very well ;-)).